### PR TITLE
 ops: epetra: revise rank-1 update

### PR DESCRIPTION
**Pull request by mzuzek**
Contents:
- [x] revised overload constraints 
- [x] added extent check
- [x] added explicit type casting for coefficients
- [x] fixed NaN injection through zero coefficients and added relevant unit test

refs #448

### CODE REVIEW

**Comment by mzuzek** [comment]
@fnrizzi 

Here in Epetra all coefficients have the same `scalar_t` type.
Would you prefer to allow different coefficient types `a_Type`, `b_Type`, etc. like in [Eigen](https://github.com/Pressio/pressio/blob/develop/include/pressio/ops/eigen/ops_rank1_update.hpp#L190-L194) ?

**Comment by mzuzek** [comment]
@fnrizzi 

Here in Epetra all vectors have the same `T` type.
Would you prefer to allow different vector types `T1`, `T2`, etc. like in [Eigen](https://github.com/Pressio/pressio/blob/develop/include/pressio/ops/eigen/ops_rank1_update.hpp#L190-L194) ?

**Comment by fnrizzi** [comment]
> @fnrizzi
> 
> Here in Epetra all coefficients have the same `scalar_t` type. Would you prefer to allow different coefficient types `a_Type`, `b_Type`, etc. like in [Eigen](https://github.com/Pressio/pressio/blob/develop/include/pressio/ops/eigen/ops_rank1_update.hpp#L190-L194) ?

yes please!

**Comment by fnrizzi** [comment]
> 

yes, we should have separate types T1, T2, etc

File: tests/functional_small/ops/ops_epetra_vector.cc [hunk 1]
```diff
@@ -297,7 +300,86 @@ TEST_F(epetraVectorGlobSize15Fixture, vector_update4_b)
...
+    }
+}
+
+TEST_F(ops_epetra, vector_update_nan)
+{
+    auto v = pressio::ops::clone(*myVector_);
+    pressio::ops::fill(v, 1.);
+    auto a = pressio::ops::clone(*myVector_);
+    pressio::ops::fill(a, 1.);
+    auto nan = pressio::ops::clone(*myVector_);
```
Comment (fnrizzi) [inline tests/functional_small/ops/ops_epetra_vector.cc:322]: `nan -> vecOfNan`

**Review by fnrizzi — CHANGES_REQUESTED** [review]
see comment

File: tests/functional_small/ops/ops_epetra_vector.cc (hunk 1 above)
Comment (mzuzek) [reply tests/functional_small/ops/ops_epetra_vector.cc]: fixed on 8da25625f132b7bb87b2d1d6423961a8826d2078

**Review by fnrizzi — APPROVED** [review]

**Comment by mzuzek** [comment]
thanks @mzuzek for your review and comments!

**Pull request merged**